google-cloud-pubsub/lib/google/cloud/pubsub/subscriber/stream.rb

```diff
@@ -0,0 +1,375 @@
+
+              @push_thread_pool.shutdown
+              @push_thread_pool.wait_for_termination
+            end
+
+            self
+          end
+
+          ##
+          # @private
+          def acknowledge *messages
+            ack_ids = coerce_ack_ids messages
+            return true if ack_ids.empty?
+
+            async_acknowledger # To make sure the object is loaded
+
+            synchronize do
+              @async_acknowledger.acknowledge ack_ids
+              @inventory.remove ack_ids
+              unpause_streaming!
+            end
+
+            true
+          end
+
+          ##
+          # @private
+          def delay deadline, *messages
+            mod_ack_ids = coerce_ack_ids messages
+            return true if mod_ack_ids.empty?
+
+            async_delayer # To make sure the object is loaded
+
+            synchronize do
+              @async_delayer.delay deadline, mod_ack_ids
+              @inventory.remove mod_ack_ids
+              unpause_streaming!
+            end
+
+            true
+          end
+
+          def async_acknowledger
+            synchronize { @async_acknowledger ||= AsyncAcknowledger.new self }
+          end
+
+          def async_delayer
+            synchronize { @async_delayer ||= AsyncDelayer.new self }
+          end
+
+          def push request
+            synchronize { @request_queue.push request }
+          end
+
+          def inventory
+            synchronize { @inventory }
+          end
+
+          ##
+          # @private
+          def delay_inventory!
+            async_delayer # To make sure the object is loaded
+
+            synchronize do
+              return true if @inventory.empty?
+
+              @async_delayer.delay subscriber.deadline, @inventory.ack_ids
+            end
+
+            true
+          end
+
+          # @private
+          def to_s
+            format "(inventory: %i)", inventory.count
+          end
+
+          # @private
+          def inspect
+            "#<#{self.class.name} #{self}>"
+          end
+
+          protected
+
+          # rubocop:disable all
+
+          def background_run enum
+            until synchronize { @stopped }
+              synchronize do
+                if @paused
+                  @pause_cond.wait
+                  next
+                end
+              end
+
+              begin
+                # Cannot syncronize the enumerator, causes deadlock
+                response = enum.next
+                response.received_messages.each do |rec_msg_grpc|
+                  rec_msg = ReceivedMessage.from_grpc(rec_msg_grpc, self)
+                  synchronize do
+                    @inventory.add rec_msg.ack_id
+
+                    perform_callback_async rec_msg
+                  end
+                end
+                synchronize { pause_streaming! }
+              rescue StopIteration
+                break
+              end
+            end
+          rescue GRPC::DeadlineExceeded
```

Comment:
This rescue is specifically to work around a bug in the GAPIC code, where the GAPIC code kills the stream when reaching the timeout. So this rescue here is only a workaround and not part of the standard error handling for API calls.

The UNAVAILABLE and CANCELLED errors are already properly handled with retry and backoff in the GAPIC code when creating the stream by calling the StreamingPull. My understanding is that the code using the stream should not be retrying with backoff _on_ the stream, since the stream is already created.

@mdietz94 Is it be acceptable to handle UNAVAILABLE and CANCELLED errors on the stream object by recreating the stream?